File: hunts/2026-07-11-5/phase-0/agents/buyer-reality/collect.py
from bs4 import BeautifulSoup
from pathlib import Path
from urllib.request import Request, urlopen
from urllib.parse import urljoin, urlparse
from datetime import datetime, timezone
import json, re, html

ROOT=Path(__file__).resolve().parent
ROOT.mkdir(parents=True,exist_ok=True)
NOW=datetime.now(timezone.utc).isoformat()
records=[]; queries=[]; failures=[]

def fetch(url, name):
    p=ROOT/name
    try:
        req=Request(url,headers={'User-Agent':'Mozilla/5.0 studio-hunt-buyer-reality/1.0'})
        with urlopen(req,timeout=30) as r: body=r.read()
        p.write_bytes(body)
        return body.decode('utf-8','replace'),p
    except Exception as e:
        failures.append({'url':url,'status':'error','error':type(e).__name__+': '+str(e),'fallbackAttempted':'none; public page was primary route','observedAt':NOW})
        return None,p

def clean(s): return ' '.join(html.unescape(s or '').split())
def add(source,url,title,desc,price,artifact,author='anonymous marketplace buyer',tools=None):
    if not title or len(desc)<60: return
    rid=f'br5-{len(records)+1:04d}'
    money='exact_job_budget' if price else 'none'
    records.append({
      'recordId':rid,'source':source,'url':url,'observedAt':NOW,'publishedAt':None,'window':'current',
      'sourceLane':'buyer_reality','retrievalStatus':'verified','actorType':'buyer_operator',
      'actorEvidence':'A buyer-authored public project brief requests paid delivery of the described operational work.',
      'statementType':'own_purchase','firsthand':True,'startupName':None,'productDomain':urlparse(url).netloc,
      'founderHandle':None,'buyer':'Marketplace client commissioning the described work','trigger':title,
      'input':None,'repeatedAction':desc[:1800],'output':title,'destination':'Delivery to the commissioning marketplace client',
      'frequency':None,'timeSpent':None,'priceOrWage':price,'commercialMetric':None,'metricPeriod':None,
      'currentTools':tools or [],'remainingManualWork':desc[:800],'objection':None,'requestedOutcome':title,
      'textExcerpt':desc[:2200]+((' Budget/rate: '+price) if price else ''),'artifactPath':str(artifact),
      'independenceKey':f'{urlparse(url).netloc}|{author}|{url.rstrip("/").split("/")[-1]}|public-marketplace',
      'commercialStrength':'medium' if price else 'weak','caveat':'Marketplace posting verifies stated purchase intent, not completed payment.',
      'classificationConfidence':'high','classificationReason':'The hydrated public listing contains a buyer-written scope and an explicit request to hire; actor identity beyond buyer role is not inferred.',
      'moneyType':money})

cats=['accounting','data-entry','intuit-quickbooks','virtual-assistant','excel','payroll','tax','legal','project-management','web-scraping']
for cat in cats:
  url=f'https://www.freelancer.com/jobs/{cat}/'
  txt,p=fetch(url,f'artifact-freelancer-{cat}.html')
  count=0
  if txt:
    s=BeautifulSoup(txt,'html.parser')
    for card in s.select('[data-project-card="true"]'):
      a=card.select_one('[data-heading-link="true"]')
      d=card.select_one('.JobSearchCard-primary-description')
      pr=card.select_one('.JobSearchCard-secondary-price')
      if not a or not d: continue
      tools=[clean(x.get_text()) for x in card.select('.JobSearchCard-primary-tagsLink')]
      add('freelancer:'+cat,urljoin(url,a.get('href')),clean(a.get_text()),clean(d.get_text()),clean(pr.get_text()) if pr else None,p,tools=tools)
      count+=1
      if count>=8: break
  queries.append({'query':cat,'source':'freelancer public category index','window':'current','resultCount':count,'purpose':'hydrate paid operational work and workaround purchases','url':url,'observedAt':NOW})

for page in [1,2,3]:
  url='https://www.peopleperhour.com/freelance-jobs/business'+(('?page='+str(page)) if page>1 else '')
  txt,p=fetch(url,f'artifact-peopleperhour-business-{page}.html')
  count=0
  if txt:
    s=BeautifulSoup(txt,'html.parser')
    seen=set()
    for a in s.find_all('a',href=True):
      href=a['href']; title=clean(a.get_text())
      if 'peopleperhour.com/freelance-jobs/business/' not in href or href in seen or not re.search(r'-\d{5,}$',href): continue
      seen.add(href)
      box=a.find_parent(['li','article','div'])
      for _ in range(3):
        if box and len(clean(box.get_text()))<100: box=box.parent
      desc=clean(box.get_text()) if box else title
      price=None
      m=re.search(r'(?:£|\$|€)\s?[\d,.]+(?:\s*(?:per hour|/hr|fixed))?',desc,re.I)
      if m: price=m.group(0)
      add('peopleperhour:business',href,title,desc[:2200],price,p)
      count+=1
      if count>=12: break
  queries.append({'query':f'business page {page}','source':'PeoplePerHour public job index','window':'current','resultCount':count,'purpose':'hydrate service purchases and operational work','url':url,'observedAt':NOW})

# Explicitly record blocked occupational-community route instead of treating discovery snippets as evidence.
failures.append({'url':'https://www.reddit.com/r/bookkeeping/search.json?q=spreadsheet&restrict_sr=on&sort=new','status':'blocked','error':'HTTP 403 Blocked','fallbackAttempted':'old.reddit.com JSON also returned HTTP 403; no snippets retained','observedAt':NOW})

# Category indexes overlap. Preserve only one normalized record per public posting.
deduped=[]; seen_keys=set()
for r in records:
  if r['independenceKey'] in seen_keys: continue
  seen_keys.add(r['independenceKey']); deduped.append(r)
records=deduped
for i,r in enumerate(records,1): r['recordId']=f'br5-{i:04d}'

for name,data in [('records.jsonl',records),('queries.jsonl',queries),('failures.jsonl',failures)]:
  (ROOT/name).write_text(''.join(json.dumps(x,ensure_ascii=False)+'\n' for x in data))

def tally(field):
 out={}
 for r in records: out[r.get(field)]=out.get(r.get(field),0)+1
 return out
missing={k:sum(1 for r in records if r.get(k) in (None,'',[])) for k in ['buyer','trigger','input','repeatedAction','output','destination','frequency','timeSpent','priceOrWage','currentTools','remainingManualWork']}
metrics={'observedAt':NOW,'totalRecords':len(records),'bySource':tally('source'),'byWindow':tally('window'),'byActor':tally('actorType'),'byStatement':tally('statementType'),'byVerification':tally('retrievalStatus'),'byClassificationConfidence':tally('classificationConfidence'),'byMoneyType':tally('moneyType'),'exactJobMoney':sum(r['moneyType']=='exact_job_budget' for r in records),'firsthand':sum(r['firsthand'] for r in records),'uniqueIndependenceKeys':len(set(r['independenceKey'] for r in records)),'uniqueHostDomains':len(set(urlparse(r['url']).netloc for r in records)),'missingFields':missing,'queryCount':len(queries),'failureCount':len(failures),'suspectedDuplicates':len(records)-len(set(r['independenceKey'] for r in records)),'adaptiveStopping':{'batchDefinition':'each independent marketplace category/page','consecutiveSaturatedBatches':0,'stoppedBecause':'assigned lane time ceiling after source rotation; additional categories remained productive'}}
(ROOT/'metrics.json').write_text(json.dumps(metrics,indent=2,ensure_ascii=False)+'\n')
print(json.dumps(metrics,indent=2))

File: hunts/2026-07-11-4/phase-0/agents/buyer-reality/collect.mjs
import fs from 'node:fs/promises';
import path from 'node:path';

const out = path.dirname(new URL(import.meta.url).pathname);
const observedAt = new Date().toISOString();
const records = [], queries = [], failures = [];
let seq = 0;
const sleep = ms => new Promise(r => setTimeout(r, ms));
const clean = s => (s || '').replace(/<[^>]+>/g, ' ').replace(/&quot;/g, '"').replace(/&#39;|&apos;/g, "'").replace(/&amp;/g, '&').replace(/&lt;/g, '<').replace(/&gt;/g, '>').replace(/\s+/g, ' ').trim();
const moneyRe = /(?:[$£€]\s?\d[\d,.]*(?:\s?(?:k|m))?|\d[\d,.]*\s?(?:usd|eur|gbp|dollars?|pounds?|euros?))/i;
const workflowRe = /\b(?:manually|spreadsheet|excel|every day|daily|every week|weekly|every month|monthly|each time|each client|each order|repeat(?:ed|ing)?|workflow|copy(?:ing)? and paste|export|import|reconcil|track(?:ing)?|schedule|invoice|report(?:ing)?|submit(?:ting)?|review(?:ing)?|check(?:ing)?|upload(?:ing)?|download(?:ing)?)\b/i;
const ownWorkflowRe = /\b(?:I|we|my|our)\b[^.!?]{0,160}\b(?:manually|use|using|run|manage|handle|track|schedule|invoice|report|submit|review|check|upload|download|export|import|reconcil|copy|pay|paid)\b/i;
const exactSpendContextRe = /\b(?:pay|paid|charge|charged|cost|costs|budget|hire|hired|wage|salary|rate|fee|expense|invoice)\b/i;
const disqualifyingMoneyRe = /\b(?:revenue|valuation|tax rate|workforce|shares?|stock|funding|raise|asking price|mrr|arr)\b/i;

function windowFor(publishedAt) {
  const ageDays = (Date.now() - new Date(publishedAt).getTime()) / 86400000;
  if (!Number.isFinite(ageDays)) return 'unknown';
  if (ageDays <= 30) return 'recent-30d';
  if (ageDays <= 90) return 'recent-90d';
  if (ageDays <= 365) return 'recent-365d';
  return 'historical-over-365d';
}

function moneyType(text, amount, hasWorkflow) {
  if (!amount) return 'none';
  if (disqualifyingMoneyRe.test(text)) return 'revenue_or_valuation';
  if (/\b(?:suppose|imagine|example|let'?s say|making up numbers)\b/i.test(text)) return 'hypothetical';
  if (hasWorkflow && exactSpendContextRe.test(text)) return /\b(?:wage|salary|hourly rate|hire|hired)\b/i.test(text) ? 'exact_job_wage' : 'exact_job_spend';
  return exactSpendContextRe.test(text) ? 'adjacent_spend' : 'currency_mention';
}

async function getJson(url, source, purpose, window = 'current') {
  const started = Date.now();
  try {
    const res = await fetch(url, {headers: {'User-Agent': 'studio-hunt-buyer-reality/1.0'}});
    if (!res.ok) throw new Error(`HTTP ${res.status}`);
    const body = await res.json();
    queries.push({query: url.replace(/([?&](?:key|token)=[^&]+)/gi, ''), source, window, resultCount: Array.isArray(body?.items) ? body.items.length : Array.isArray(body?.hits) ? body.hits.length : null, purpose, observedAt, elapsedMs: Date.now()-started});
    return body;
  } catch (e) {
    failures.push({source, url, window, status: 'error', error: String(e), fallbackAttempted: null, observedAt});
    return null;
  }
}

function add({source,url,publishedAt,actorType='unknown',actorEvidence=null,statementType,firsthand,buyer=null,trigger=null,input=null,repeatedAction=null,output=null,destination=null,frequency=null,timeSpent=null,priceOrWage=null,moneyType='none',currentTools=[],remainingManualWork=null,requestedOutcome=null,textExcerpt,artifactPath,caveat=null,author='unknown',dataset, productDomain=null, classificationConfidence='low', classificationReason=null}) {
  records.push({recordId:`br-${String(++seq).padStart(4,'0')}`,source,url,observedAt,publishedAt,window:windowFor(publishedAt),sourceLane:'buyer_reality',retrievalStatus:'verified',actorType,actorEvidence,statementType,firsthand,startupName:null,productDomain,founderHandle:null,buyer,trigger,input,repeatedAction,output,destination,frequency,timeSpent,priceOrWage,moneyType,commercialMetric:null,metricPeriod:null,currentTools,remainingManualWork,objection:null,requestedOutcome,textExcerpt,artifactPath,independenceKey:`${new URL(url).hostname}|${author}|none|${dataset}`,commercialStrength:['exact_job_spend','exact_job_wage'].includes(moneyType)?'medium':'none',classificationConfidence,classificationReason,caveat});
}

const sites = ['workplace','freelancing','money','academia','travel','diy','webmasters','salesforce','magento','wordpress','craftcms','law','patents','graphicdesign','writers','projectmanagement','gardening','homebrewing'];
for (const site of sites) {
  const url = `https://api.stackexchange.com/2.3/questions?site=${site}&pagesize=60&order=desc&sort=activity&filter=withbody`;
  const body = await getJson(url, `stackexchange:${site}`, 'Broad occupational question traversal; retain own workflows and exact-money observations');
  if (!body) continue;
  await fs.writeFile(path.join(out,`artifact-stackexchange-${site}.json`), JSON.stringify(body));
  let kept = 0;
  for (const q of body.items || []) {
    const text = clean(`${q.title}. ${q.body}`);
    if (!(workflowRe.test(text) || moneyRe.test(text))) continue;
    const exactMoney = text.match(moneyRe)?.[0] || null;
    const workflow = workflowRe.test(text);
    const firstPerson = ownWorkflowRe.test(text);
    const tools = [...new Set((text.match(/\b(?:Excel|Google Sheets|spreadsheet|Salesforce|WordPress|Magento|QuickBooks|Notion|Airtable|Jira|email|PDF)\b/gi)||[]).map(x=>x.toLowerCase()))];
    add({source:`stackexchange:${site}`,url:q.link,publishedAt:new Date(q.creation_date*1000).toISOString(),actorType:'unknown',actorEvidence:`Self-authored question on ${site}; occupation not inferred from community membership.`,statementType:firstPerson&&workflow?'own_workflow':'unknown',firsthand:firstPerson&&workflow,buyer:null,trigger:clean(q.title),input:null,repeatedAction:firstPerson&&workflow?text:null,output:null,destination:null,frequency:(text.match(/\b(?:daily|every day|weekly|every week|monthly|every month|each time|each client|each order)\b/i)||[])[0]||null,timeSpent:(text.match(/\b\d+(?:\.\d+)?\s*(?:minutes?|hours?|days?|weeks?)\b/i)||[])[0]||null,priceOrWage:exactMoney,moneyType:moneyType(text,exactMoney,firstPerson&&workflow),currentTools:tools,remainingManualWork:/manually/i.test(text)?text:null,requestedOutcome:q.title,textExcerpt:text,artifactPath:`${out}/artifact-stackexchange-${site}.json`,author:String(q.owner?.user_id||q.owner?.display_name||'unknown'),dataset:`stackexchange-${site}`,productDomain:q.link?new URL(q.link).hostname:null,classificationConfidence:'low',classificationReason:'Regex-selected first-person workflow lead; requires parent quote audit.',caveat:'Community membership does not establish buyer role; question title and forum destination are not job-sentence fields.'});
    if (++kept >= 8) break;
  }
  await sleep(120);
}

const hnTerms = ['"I pay"','"we pay"','"I paid"','"costs us"','"per month"','"per hour"','manually','spreadsheet','"every week"','"each month"'];
for (const term of hnTerms) {
  const url = `https://hn.algolia.com/api/v1/search_by_date?tags=ask_hn&hitsPerPage=100&query=${encodeURIComponent(term)}`;
  const body = await getJson(url,'hn_algolia_ask','Broad Ask HN buyer language discovery and hydration');
  if (!body) continue;
  const slug = term.replace(/[^a-z0-9]+/gi,'-').replace(/^-|-$/g,'').toLowerCase();
  await fs.writeFile(path.join(out,`artifact-hn-${slug}.json`), JSON.stringify(body));
  let kept=0;
  for (const h of body.hits || []) {
    const text=clean(`${h.title||''}. ${h.story_text||h.comment_text||''}`);
    if (text.length<80 || !/\b(?:I|we|my|our)\b/i.test(text) || !(workflowRe.test(text)||moneyRe.test(text))) continue;
    const exactMoney=text.match(moneyRe)?.[0]||null;
    const workflow=workflowRe.test(text);
    const id=h.objectID;
    const ownWorkflow=workflow&&ownWorkflowRe.test(text);
    add({source:'hn_algolia_ask',url:`https://news.ycombinator.com/item?id=${id}`,publishedAt:h.created_at,actorType:'unknown',actorEvidence:'Ask HN author; role not explicitly established in extracted text.',statementType:ownWorkflow?'own_workflow':'unknown',firsthand:ownWorkflow,trigger:h.title||null,repeatedAction:ownWorkflow?text:null,output:null,destination:null,frequency:(text.match(/\b(?:daily|every day|weekly|every week|monthly|every month|each time|each client|each order)\b/i)||[])[0]||null,timeSpent:(text.match(/\b\d+(?:\.\d+)?\s*(?:minutes?|hours?|days?|weeks?)\b/i)||[])[0]||null,priceOrWage:exactMoney,moneyType:moneyType(text,exactMoney,ownWorkflow),currentTools:[...new Set((text.match(/\b(?:Excel|Google Sheets|spreadsheet|Salesforce|QuickBooks|Notion|Airtable|Jira|email|PDF)\b/gi)||[]).map(x=>x.toLowerCase()))],remainingManualWork:/manually/i.test(text)?text:null,requestedOutcome:h.title||null,textExcerpt:text,artifactPath:`${out}/artifact-hn-${slug}.json`,author:h.author||'unknown',dataset:'hn-ask',productDomain:'news.ycombinator.com',classificationConfidence:'low',classificationReason:'Regex-selected Ask HN workflow lead; requires parent quote audit.',caveat:'HN audience is developer-heavy; question title and HN destination are not job-sentence output/destination.'});
    if(++kept>=5) break;
  }
  await sleep(120);
}

// Prefer the most useful evidence if APIs returned more than the lane floor.
const uniq = [...new Map(records.map(r=>[`${r.url}|${r.textExcerpt}`,r])).values()];
await fs.writeFile(path.join(out,'records.jsonl'), uniq.map(x=>JSON.stringify(x)).join('\n')+'\n');
await fs.writeFile(path.join(out,'queries.jsonl'), queries.map(x=>JSON.stringify(x)).join('\n')+'\n');
await fs.writeFile(path.join(out,'failures.jsonl'), failures.map(x=>JSON.stringify(x)).join('\n')+(failures.length?'\n':''));
const countBy = (k) => Object.fromEntries([...new Set(uniq.map(x=>x[k]))].map(v=>[String(v),uniq.filter(x=>x[k]===v).length]));
const metrics={observedAt,totalRecords:uniq.length,retrievedVerified:uniq.filter(x=>x.retrievalStatus==='verified').length,currencyMentions:uniq.filter(x=>x.priceOrWage).length,qualifiedExactJobMoney:uniq.filter(x=>['exact_job_spend','exact_job_wage'].includes(x.moneyType)&&x.classificationConfidence!=='low').length,heuristicFirsthandLeads:uniq.filter(x=>x.firsthand).length,qualifiedFirsthand:uniq.filter(x=>x.firsthand&&x.classificationConfidence!=='low').length,qualifiedRepeatedWorkflow:uniq.filter(x=>x.firsthand&&x.classificationConfidence!=='low'&&x.frequency).length,completeJobSentences:uniq.filter(x=>['buyer','trigger','input','repeatedAction','output','destination'].every(k=>x[k])).length,uniqueIndependenceKeys:new Set(uniq.map(x=>x.independenceKey)).size,uniqueSourceLabels:new Set(uniq.map(x=>x.source)).size,uniqueHostDomains:new Set(uniq.map(x=>new URL(x.url).hostname)).size,bySource:countBy('source'),byWindow:countBy('window'),byActor:countBy('actorType'),byStatement:countBy('statementType'),byVerification:countBy('retrievalStatus'),missingFields:Object.fromEntries(['buyer','trigger','input','repeatedAction','output','destination','frequency','timeSpent','priceOrWage','currentTools','remainingManualWork'].map(k=>[k,uniq.filter(x=>x[k]==null||(Array.isArray(x[k])&&x[k].length===0)).length])),queryCount:queries.length,failureCount:failures.length,suspectedDuplicates:uniq.length-new Set(uniq.map(x=>x.url)).size};
await fs.writeFile(path.join(out,'metrics.json'),JSON.stringify(metrics,null,2)+'\n');
console.log(JSON.stringify(metrics));
